store/postgres/src/writable.rs

```diff
@@ -325,37 +351,471 @@ impl WritableStore {
+/// queued
+enum Request {
+    Write {
+        /// The block at which we are writing the changes
+        block_ptr: BlockPtr,
+        firehose_cursor: Option<String>,
+        mods: Vec<EntityModification>,
+        data_sources: Vec<StoredDynamicDataSource>,
+        deterministic_errors: Vec<SubgraphError>,
+    },
+    Revert {
+        /// The subgraph head will be at this block pointer after the revert
+        block_ptr: BlockPtr,
+        firehose_cursor: Option<String>,
+    },
+}
+
+/// A queue that asynchronously writes requests queued with `push` to the
+/// underlying store and allows retrieving information that is a combination
+/// of queued changes and changes already committed to the store.
+struct Queue {
+    store: Arc<SyncStore>,
+    /// A queue of pending requests. New requests are appended at the
+    /// back, and popped off the front for processing. That means that
+    /// going front-to-back, block numbers in the requests are
+    /// increasing.
+    queue: BoundedQueue<Arc<Request>>,
+
+    /// The write task puts errors from `transact_block_operations` here so
+    /// we can report them on the next call to transact block operations.
+    write_err: Mutex<Option<StoreError>>,
+
+    /// True if the background worker ever encountered an error. Once that
+    /// happens, no more changes will be written, and any attempt to write
+    /// or revert will result in an error
+    poisoned: AtomicBool,
 }
 
-#[allow(dead_code)]
-pub struct WritableAgent {
-    store: Arc<WritableStore>,
+impl Queue {
+    /// Create a new queue and spawn a task that processes write requests
+    fn start(
+        logger: Logger,
+        store: Arc<SyncStore>,
+        capacity: usize,
+        registry: Arc<dyn MetricsRegistry>,
+    ) -> Arc<Self> {
+        async fn start_writer(queue: Arc<Queue>, stopwatch: StopwatchMetrics) {
+            loop {
+                let res = match queue.queue.peek().await.as_ref() {
+                    Request::Write {
+                        block_ptr: block_ptr_to,
+                        firehose_cursor,
+                        mods,
+                        data_sources,
+                        deterministic_errors,
+                    } => queue.store.transact_block_operations(
+                        block_ptr_to,
+                        firehose_cursor.as_deref(),
+                        mods,
+                        stopwatch.cheap_clone(),
+                        data_sources,
+                        deterministic_errors,
+                    ),
+                    Request::Revert {
+                        block_ptr,
+                        firehose_cursor,
+                    } => queue
+                        .store
+                        .revert_block_operations(block_ptr.clone(), firehose_cursor.as_deref()),
+                };
+                // Remove the tombstone that take_front left in place
+                queue.queue.pop().await;
```

Comment:
Yes, that is indeed a bit tricky. I just added this comment to the loop:
```rust
                // We peek at the front of the queue, rather than pop it
                // right away, so that query methods like `get` have access
                // to the data while it is being written. If we popped here,
                // these methods would not be able to see that data until
                // the write transaction commits, causing them to return
                // incorrect results.
```

Reply:
Thanks, that really is subtle.

> // Remove the tombstone that take_front left in place

This comment is still cryptic to me.

Reply:
I changed it to
> // The request has been handled. It's now safe to remove it
> // from the queue